drivers/wifi/esp_hosted/esp_hosted_hal.c

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (c) 2025 Arduino SA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdint.h>
+#include <string.h>
+
+#include <zephyr/kernel.h>
+#include "esp_hosted_wifi.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(esp_hosted_hal, CONFIG_WIFI_LOG_LEVEL);
+
+int esp_hosted_hal_init(const struct device *dev)
+{
+	const esp_hosted_config_t *config = dev->config;
+
+	if (!spi_is_ready_dt(&config->spi_bus)) {
+		LOG_ERR("SPI device is not ready");
+		return -ENODEV;
+	}
+
+	/* Configure pins. */
+	gpio_pin_configure_dt(&config->dataready_gpio, GPIO_OUTPUT);
+	gpio_pin_configure_dt(&config->reset_gpio, GPIO_OUTPUT);
+
+	/* Perform a hard-reset */
+	gpio_pin_set_dt(&config->dataready_gpio, 1);
+	gpio_pin_set_dt(&config->reset_gpio, 0);
+	k_msleep(100);
+	gpio_pin_set_dt(&config->reset_gpio, 1);
+	k_msleep(500);
+
+	/* Configure handshake/dataready pins. */
+	gpio_pin_configure_dt(&config->dataready_gpio, GPIO_INPUT);
+	gpio_pin_configure_dt(&config->handshake_gpio, GPIO_INPUT);
+
+	return 0;
+}
+
+bool esp_hosted_hal_data_ready(const struct device *dev)
+{
+	const esp_hosted_config_t *config = dev->config;
+
+	return gpio_pin_get_dt(&config->dataready_gpio);
+}
+
+int esp_hosted_hal_spi_transfer(const struct device *dev, void *tx, void *rx, uint32_t size)
+{
+	int ret = 0;
+	esp_hosted_data_t *data = dev->data;
+	const esp_hosted_config_t *config = dev->config;
+
+	const struct spi_buf tx_buf = {.buf = tx ? tx : rx, .len = size};
+	const struct spi_buf_set tx_set = {.buffers = &tx_buf, .count = 1};
+
+	const struct spi_buf rx_buf = {.buf = rx ? rx : tx, .len = size};
+	const struct spi_buf_set rx_set = {.buffers = &rx_buf, .count = 1};
+
+	/* Wait for handshake pin to go high. */
+	for (uint64_t start = k_uptime_get();; k_msleep(1)) {
```

Comment:
Have you considered using interrupt+semaphore here?

Reply:
I did, in another driver. I configured the pin as IRQ and notified the event handling thread, but the IRQ wasn't very reliable on its own, so I ended up using both poll + IRQ. The IRQ doesn't seem to improve latency by much, if anyway, so I decided to ignore it this time as I'm not sure it's worth it.
One thing I could improve here is to make the poll rate configurable.